File: external_apps/mobile_ios/mobile_ios/Core/Auth/TokenManager.swift
//
//  TokenManager.swift
//  mobile_ios
//
//  Created for Authentication Feature
//

import Foundation

class TokenManager {
    static let shared = TokenManager()
    
    private let defaults = UserDefaults.standard
    private let accessTokenKey = "access_token"
    private let refreshTokenKey = "refresh_token"
    private let tokenExpiryKey = "token_expiry"
    private let userIdKey = "user_id"
    
    private init() {}
    
    func saveTokens(accessToken: String, refreshToken: String, expiresIn: TimeInterval = 3600) {
        let expiryTime = Date().addingTimeInterval(expiresIn)
        defaults.set(accessToken, forKey: accessTokenKey)
        defaults.set(refreshToken, forKey: refreshTokenKey)
        defaults.set(expiryTime, forKey: tokenExpiryKey)
    }
    
    func getAccessToken() -> String? {
        return defaults.string(forKey: accessTokenKey)
    }
    
    func getRefreshToken() -> String? {
        return defaults.string(forKey: refreshTokenKey)
    }
    
    func isTokenExpired() -> Bool {
        guard let expiryTime = defaults.object(forKey: tokenExpiryKey) as? Date else {
            return true
        }
        return Date() >= expiryTime
    }
    
    func hasValidToken() -> Bool {
        guard let token = getAccessToken(), !token.isEmpty else {
            return false
        }
        return !isTokenExpired()
    }
    
    func clearTokens() {
        defaults.removeObject(forKey: accessTokenKey)
        defaults.removeObject(forKey: refreshTokenKey)
        defaults.removeObject(forKey: tokenExpiryKey)
        defaults.removeObject(forKey: userIdKey)
    }
    
    func saveUserId(_ userId: String) {
        defaults.set(userId, forKey: userIdKey)
    }
    
    func getUserId() -> String? {
        return defaults.string(forKey: userIdKey)
    }
    
    func shouldRefreshToken() -> Bool {
        guard let expiryTime = defaults.object(forKey: tokenExpiryKey) as? Date else {
            return false
        }
        let fiveMinutesFromNow = Date().addingTimeInterval(5 * 60)
        return fiveMinutesFromNow >= expiryTime
    }
    
    func getTimeUntilExpiry() -> TimeInterval {
        guard let expiryTime = defaults.object(forKey: tokenExpiryKey) as? Date else {
            return 0
        }
        let timeLeft = expiryTime.timeIntervalSinceNow
        return max(0, timeLeft)
    }
}

File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/data/auth/TokenManager.kt
package com.example.mobile_android.data.auth

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TokenManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    private val _isAuthenticated = MutableStateFlow(hasValidToken())
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated
    
    companion object {
        private const val PREFS_NAME = "auth_prefs"
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
        private const val KEY_TOKEN_EXPIRY = "token_expiry"
        private const val KEY_USER_ID = "user_id"
    }
    
    fun saveTokens(accessToken: String, refreshToken: String, expiresIn: Long = 3600) {
        val expiryTime = System.currentTimeMillis() + (expiresIn * 1000)
        prefs.edit().apply {
            putString(KEY_ACCESS_TOKEN, accessToken)
            putString(KEY_REFRESH_TOKEN, refreshToken)
            putLong(KEY_TOKEN_EXPIRY, expiryTime)
            apply()
        }
        _isAuthenticated.value = true
    }
    
    fun getAccessToken(): String? {
        return prefs.getString(KEY_ACCESS_TOKEN, null)
    }
    
    fun getRefreshToken(): String? {
        return prefs.getString(KEY_REFRESH_TOKEN, null)
    }
    
    fun isTokenExpired(): Boolean {
        val expiryTime = prefs.getLong(KEY_TOKEN_EXPIRY, 0)
        return System.currentTimeMillis() >= expiryTime
    }
    
    fun hasValidToken(): Boolean {
        val token = getAccessToken()
        return !token.isNullOrEmpty() && !isTokenExpired()
    }
    
    fun clearTokens() {
        prefs.edit().apply {
            remove(KEY_ACCESS_TOKEN)
            remove(KEY_REFRESH_TOKEN)
            remove(KEY_TOKEN_EXPIRY)
            remove(KEY_USER_ID)
            apply()
        }
        _isAuthenticated.value = false
    }
    
    fun saveUserId(userId: String) {
        prefs.edit().putString(KEY_USER_ID, userId).apply()
    }
    
    fun getUserId(): String? {
        return prefs.getString(KEY_USER_ID, null)
    }
    
    /**
     * Check if token needs refresh (expires in less than 5 minutes)
     */
    fun shouldRefreshToken(): Boolean {
        val expiryTime = prefs.getLong(KEY_TOKEN_EXPIRY, 0)
        val fiveMinutesFromNow = System.currentTimeMillis() + (5 * 60 * 1000)
        return fiveMinutesFromNow >= expiryTime
    }
    
    /**
     * Get time until token expires in seconds
     */
    fun getTimeUntilExpiry(): Long {
        val expiryTime = prefs.getLong(KEY_TOKEN_EXPIRY, 0)
        val timeLeft = expiryTime - System.currentTimeMillis()
        return if (timeLeft > 0) timeLeft / 1000 else 0
    }
}
